src/python/pants/core/util_rules/config_files.py

```diff
@@ -83,5 +87,70 @@ async def find_config_file(request: ConfigFilesRequest) -> ConfigFiles:
     return ConfigFiles(config_snapshot)
 
 
+class OrphanFilepathBehavior(Enum):
+    IGNORE = "ignore"
+    ERROR = "error"
+
+
+@dataclass(frozen=True)
+class GatheredConfigFilesByDirectories:
```

Comment:
Very nice refactor! Please split it out to a separate prerequisite PR to make this PR more manageable for reviewers.